CountDownTimer and SecondView Alpha Checker for better Performance added

rencsaridogan opened the pull request:
Hello,

I've been using DraggablePanel with ExoPlayer but there were problems since I was using SENSOR when DraggablePanel was active, and I was using PORTRAIT on minimized. I needed to check SecondView Alpha and I was in need of Timers to make the isMaximized, isMinimized better. So I added the listeners below and improved when they are triggered:

onTouchListener
clickedToMaximized
clickedToMinimize
smoothSlide

The improvements also helped me to collect true data about how much DraggablePanel is used because it was throwing events way more since onMaximized and onMinimezed were called more than once in some cases.

You have to initialize timers when you are initializing the DraggableView with the code below:

`draggable_view.initializeTimers();`

My code for the Listener looks like below:

```
draggable_view.setDraggableListener(new DraggableListener() {
            @Override
            public void onMaximized() {
                if (!draggable_view.isClickedMinimize()) {
                    if (videoPlayer != null && prerollShown)
                        videoPlayer.showPrerollLayout();
                    if (draggable_view.isMinimized())
                        functions.sendEventAnalytics("panel", "maximized");
                    videoPlayer.setVisibility(View.VISIBLE);
                    setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_SENSOR);
                } else {
                    draggable_view.setClickedtoMinimize();
                }
            }

            @Override
            public void onMinimized() {
                if (!draggable_view.isClickedMaximize()) {
                    if (videoPlayer != null && prerollShown)
                        videoPlayer.hidePrerollLayout();
                    functions.sendEventAnalytics("panel", "minimized");
                    functions.colorComponentsNormal(MainActivity.this);
                    if (draggable_view.getSecondViewAlpha() < 0.15) {
                        new Handler().postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                videoPlayer.setVisibility(View.VISIBLE);
                                setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
                            }
                        }, 300);
                    }
                } else {
                    draggable_view.setClickedtoMaximize();
                }
            }

            @Override
            public void onClosedToLeft() {
                getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
                functions.sendEventAnalytics("panel", "closed_to_left");
                draggableViewGone = true;
                if (videoPlayer != null)
                    if (videoPlayer.isPlaying())
                        videoPlayer.pause();
            }

            @Override
            public void onClosedToRight() {
                getWindow().clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
                functions.sendEventAnalytics("panel", "closed_to_right");
                draggableViewGone = true;
                if (videoPlayer != null)
                    if (videoPlayer.isPlaying())
                        videoPlayer.pause();
            }

            @Override
            public void onTouchListener() {
                draggable_view.setSeekBarStatus(videoPlayer.areContolsVisible());
                if (draggable_view.getSecondViewAlpha() < 0.99 && !draggable_view.isMinimized()) {
                    setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
                } else if (draggable_view.isMinimized() && draggable_view.getSecondViewAlpha() > 0.1) {
                    setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
                }
            }

            @Override
            public void clickedToMinimize() {
            }

            @Override
            public void smoothSlide() {
            }

            @Override
            public void clickedToMaximize() {
            }

        });
```

Best regards,
Renç Sarıdoğan.

pedrovgs commented:
Hi @rencsaridogan, thanks for your PR. I'm going to review it right now.

I've noticed that your PR is breaking the checkstyle rules I've configured. Can you review the code and format it to match with the checkstyle? If you want to do it automatically you can use this project to configure your IDE: https://github.com/square/java-code-styles

pedrovgs commented on draggablepanel/src/main/java/com/github/pedrovgs/DraggableListener.java:
```diff
@@ -42,4 +42,28 @@
+   * Called when the view is being dragged to either minimize or maximize
+   */
+
+  void smoothSlide();
```
We should change the naming of the methods you've added to match with the naming of the project. If you review the methods already implemented all start using `on` as prefix. Based on that, we should rename these last three methods to:

``` java
void onTouch()

void onClickedToMaximize()

void onClickedToMinimize();

void onSmoothSlide();
```

rencsaridogan replied:
Working on it.

pedrovgs commented on draggablepanel/src/main/java/com/github/pedrovgs/DraggableView.java:
```diff
@@ -359,6 +365,7 @@ public boolean isClosed() {
    * @return true if the touch event is realized over the drag or second view.
    */
   @Override public boolean onTouchEvent(MotionEvent ev) {
+    listener.onTouchListener();
```
It could be useful to pass the event to the listener method.

pedrovgs commented on draggablepanel/src/main/java/com/github/pedrovgs/DraggableView.java:
```diff
@@ -366,7 +373,9 @@ public boolean isClosed() {
-    viewDragHelper.processTouchEvent(ev);
+    if(getContext().getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT){
+      viewDragHelper.processTouchEvent(ev);
+    }
```
Why do you process the touch event just if the configuration is portrait?

rencsaridogan replied:
In case of usage a video player such as ExoPlayer, this processes the drag event when you try to hover the player controls, breaks the user experience about controlling the player. A control has been added for that purpose. Could be optional for people who doesn't have hover for player controls and aims to drag even on Landscape.

pedrovgs replied:
Definitively this should be configurable. You should add a `onlyPortrait` flag in to the `attrs.xml` file and use this value here.

rencsaridogan replied:
@pedrovgs I thought it is better to give all the options (all, landscape, portrait) for the user, in case of different needs I'll be adding a flag to support all of those to trigger touch events. You can check it out after the commit with new implementations.

pedrovgs commented on draggablepanel/src/main/java/com/github/pedrovgs/DraggableView.java:
```diff
@@ -260,7 +266,7 @@ public void minimize() {
   public void closeToRight() {
     if (viewDragHelper.smoothSlideViewTo(dragView, transformer.getOriginalWidth(),
-        getHeight() - transformer.getMinHeightPlusMargin())) {
+            getHeight() - transformer.getMinHeightPlusMargin())) {
```
All this extra whitespace you have here and in the rest of the file will be gone if you format the project using the correct checkstyle.

rencsaridogan replied:
I'll be formatting the code with the correct checkstyle along with the other changes.

pedrovgs replied:
Cool, thanks!

pedrovgs commented on draggablepanel/src/main/java/com/github/pedrovgs/DraggableView.java:
```diff
@@ -784,4 +798,73 @@ private void notifyCloseToLeftListener() {
+      public void onFinish() {
+
+      }
+    };
```
@rencsaridogan this is not the way to go to implement the listener mechanism. You are using a polling mechanism to notify your listener and depending on the device this approach will notify at the second X or X+1. You should notify your listener close to the method where the alpha is modified and remove the usage of this count down timer.

rencsaridogan replied:
You are right. Since the alpha is modified constantly in that case, removing the extra timers and handling the notifiers within the alpha change sounds logical. I'll be implementing and testing that version. Thanks!

pedrovgs replied:
@rencsaridogan if you add an example to the project using exo player could be great. We could use it to improve the library usage combined with this component.

rencsaridogan replied:
@pedrovgs While I'm coding the changes through, I will add a sample for ExoPlayer along with others for the purpose of improving usage of both components together. I believe since industry is focused on using ExoPlayer based players, so this could be great to give it a full support!

rencsaridogan commented:
@pedrovgs I've implemented a change for anyone to be able to choose any rotation they like for the panel to work (Default is all), implemented the changes and applied the improvements we've discussed. ExoPlayer example is not added yet, working on that to prepare the best example to work on. 

After the review you can merge this without waiting the example version. After the fixes and improvements it shouldn't effect the current version anyhow. Instead it should be adding the new features.

Best regards,
Renç Sarıdoğan.

pedrovgs commented:
Sorry for the delay @rencsaridogan, I've been out for some time. I think I can wait for the ExoPlayer example before to merge this PR. Please, if you have time add the example and the documentation needed to show how to use the new feature :)
